keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+The alternative services IPs exposed are the same published in the Endpoints for the `kubernetes.default` services, the Alt-Svc headers will not be present on requests
+that are not authenticated.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problem can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 
+
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+The Alternative Services RFC defines options to [implement caching](https://datatracker.ietf.org/doc/html/rfc7838#section-3.1),
+however, it doesn't seem necessary because:
+
+- kubernetes control-plane is not a high dynamic environment.
```

Comment:
I don't think you can be sure enough about this to treat it as a given.

Reply:
there has to be always 1 IP alive, right? the etcd cluster guarantees certain stability, is not that the control-plane IPs are renumbered every x minutes XD

Reply:
etcd IPs are not necessarily apiserver IPs.

Things actually could all be down at once and the system should recover.

Reply:
@aojea 

> the etcd cluster guarantees certain stability, is not that the control-plane IPs are renumbered every x minutes XD

Sorry, I didn't catch it. Could you rephrase it? Also I'd like that I am playing around autoscaling of API servers with HPA, so the thesis about stable "control planes IPs" looks strange for me. Of course, I am providing to client just one IP of LB, which hides all API servers. I may refer to https://github.com/gardener/gardener project and to https://github.com/kvaps/kubernetes-in-kubernetes projects which allow some kind of similar setup

Reply:
You have to have always.one stable IP, you can't switch all the ips at the same time, can you?

Reply:
This model doesn't aallow you to switch *all* the IPs ever, right?  Once you configure a client  with an IP, you have to keep that IP valid.

Reply:
the bootstrap seed is always in the client configuration file, if you use IPs on the clients and you change all the IPs doesn't matter the model (we can't use broadcast as in bootp/dhcp model here, maybe metadata on cloud environments, but that is not generalizable), you always need to change the configuration UNLESS you use an URL in the client, so instead of reconfiguring the clients, you just change the DNS record ...

Reply:
If you are forced to either a) update all your clients or b) update DNS, then I'm still finding the value of alt-svc to be quite diminished.

Reply:
that is only for a big change as in "you renumber the IPs of your whole control plane"
I clarified my assumptions https://github.com/kubernetes/enhancements/pull/3034#issuecomment-1018301393 ,